### PR TITLE
Add wavefront obj and x3d export

### PR DESCRIPTION
**Pull request by tu500**
TODO:
* [ ] Code review
* [ ] Add a few test cases
* [ ] This PR exhibits the same export issues as AMF and OFF. If merged, we should create new issues tracking this for OBJ and X3D as well, referencing the two mentioned issues: #909 and #910

### CODE REVIEW

**Comment by kintel** [comment]
@donbright Didn't you explore OBJ export/import a while back? Could you take a quick look at how this compares to your explorations?

File: Info.plist [hunk 1]
```diff
@@ -93,6 +93,26 @@
...
 		</dict>
+		<dict>
+			<key>CFBundleTypeExtensions</key>
+			<array>
+				<string>x3d</string>
+			</array>
+			<key>CFBundleTypeName</key>
+			<string>X3D file</string>
+			<key>CFBundleTypeRole</key>
+			<string>Viewer</string>
```
Comment (kintel) [inline Info.plist]: We cannot import X3D files, so this entire block should be removed. Same for OBJ.

File: src/MainWindow.ui [hunk 2]
```diff
@@ -1288,6 +1290,24 @@
     <string>Export as &amp;AMF...</string>
    </property>
   </action>
+  <action name="fileActionExportX3D">
+   <property name="icon">
+    <iconset resource="../openscad.qrc">
+     <normaloff>:/images/X3D.png</normaloff>:/images/X3D.png</iconset>
```
Comment (kintel) [inline src/MainWindow.ui]: This image resource is missing. Same for `OBJ.png`

File: src/export_indexed_triangles.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
+/*
+ *	OpenSCAD (www.openscad.org)
+ *	Copyright (C) 2009-2011 Clifford Wolf <clifford@clifford.at> and
+ *													Marius Kintel <marius@kintel.net>
```
Comment (kintel) [inline src/export_indexed_triangles.cc]: Indentation glitch

File: src/export_indexed_triangles.cc [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+export_indexed_triangles(const PolySet &ps)
+{
+	PolySet triangulated(3);
+	PolysetUtils::tessellate_faces(ps, triangulated);
+
+	std::vector<export_vec3> vecs;
+	std::vector<export_indexed_triangle> tris;
+
+	for(const auto &p : triangulated.polygons) {
+		assert(p.size() == 3); // STL only allows triangles
```
Comment (kintel) [inline src/export_indexed_triangles.cc]: Comment is misleading

File: src/export_indexed_triangles.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <vector>
+#include <utility>
+
+#include "polyset.h"
+
+
+struct export_vec3 {
```
Comment (kintel) [inline src/export_indexed_triangles.h]: Why not use `Eigen::Vector3d` as we do other places, see `linalg.h`

File: src/export_indexed_triangles.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	double x, y, z;
+
+	bool operator==(const export_vec3 &rhs) const {
+		return x == rhs.x && y == rhs.y && z == rhs.z;
+	}
+};
+
+struct export_indexed_triangle {
```
Comment (kintel) [inline src/export_indexed_triangles.h]: We could probably use `IndexedTriangle` here. See `GeometryUtils.h`

File: src/export_indexed_triangles.cc [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+#include "polyset.h"
+#include "polyset-utils.h"
+
+#ifdef ENABLE_CGAL
+#include "CGAL_Nef_polyhedron.h"
+#include "cgal.h"
+#include "cgalutils.h"
+
+
+static size_t insert_vector(std::vector<export_vec3> &vecs, const export_vec3 &vec)
```
Comment (kintel) [inline src/export_indexed_triangles.cc]: This is not very performant. A better idea could be to use `Reindexer` for this; see 'Reindexer.h'.
`Reindexer` is used multiple other places in the codebase.

File: src/export_obj.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+static void append_obj(const shared_ptr<const Geometry> &geom, std::ostream &output)
+{
+	if (const CGAL_Nef_polyhedron *N = dynamic_cast<const CGAL_Nef_polyhedron *>(geom.get())) {
+		append_obj(*N, output);
+	}
+	else if (const PolySet *ps = dynamic_cast<const PolySet *>(geom.get())) {
+		auto r = export_indexed_triangles(*ps);
+		append_obj(r.first, r.second, output);
+	}
+	// else if (const Polygon2d *poly = dynamic_cast<const Polygon2d *>(geom.get())) {
```
Comment (kintel) [inline src/export_obj.cc]: Any reasons why this was commented out? It looks resonable.

File: src/MainWindow.ui (hunk 2 above)
Comment (tu500) [inline src/MainWindow.ui]: Do you have a way to generate new icons, a script? Where do the existing come from?

File: src/export_obj.cc (hunk 8 above)
Comment (tu500) [inline src/export_obj.cc]: Yes it is, I just got confused about the capabilities of obj files, sorry.

**Comment by tu500** [comment]
Considering test cases, I don't know enough about the code base to generate good test cases and I didn't find any tests for the other exports. If output is deterministic one could do something like

```sh
echo 'cube(10);' > test.scad
openscad -o test.x3d test.scad
diff test.x3d expected.x3d
```

but that doesn't seem right.

Regarding the export issues you mentioned: I do have an overload for PolySets. Do you mean something else by that?

File: src/MainWindow.ui (hunk 2 above)
Comment (kintel) [inline src/MainWindow.ui]: They are hand-drawn. There are some SVGs buried in a PR somewhere (#1214). I don't think icons are necessary unless these icons make it into a toolbar.

**Comment by kintel** [comment]
Tests: It's tricky to test without implementing the corresponding import functionality. I think our current tests rely on bidirectional export/import. A naive check assuming deterministic test output for a very simple testfile is better than no test at all. Note: All our tests are black-box regression tests, so they are pretty easy to implement. See here: https://github.com/openscad/openscad/blob/master/doc/testing.txt

**Comment by kintel** [comment]
Export issues: The challenge is that these exports, as far as I can tell, convert directly from a Nef_3 polyhedron into a CGAL Polyhedron, then into the correct file format. This conversion fails in some corner cases. If you look at the e.g. STL export vs. AMF export, you can see that the code path is quite different. The mentioned issues are tracking this discrepancy. I'm not expecting you to fix that, but we should add issues so we know to deal with it later.

..of course, if you decide to fix it that would be awesome :)

**Comment by tu500** [comment]
Ok, this might be getting off topic. As far as I understand it the STL code converts Nef polyhedrons to PolySets and exports those. There is also a code path which converts Nef polyhedrons to CGAL polyhedrons and exports those, but this one is unused. The AMF code converts PolySets to Nef polyhedrons, those to CGAL polyhedrons and then exports them.

The point I wanted to make was that I copied the STL workflow.

**Comment by tu500** [comment]
Also, thanks for the review and especially for explicitly pointing to the right headers. I redid the code and addressed your comments. I'll look into test code and icons.

**Comment by MichaelAtOz** [comment]
@tu500 What did you create the icons with?

**Comment by tu500** [comment]
@MichaelAtOz As @kintel mentioned there is a PR with SVG icons #1214.

**Comment by tu500** [comment]
I should mention I added these icons, together with a conversion script, into my svg-icons branch. @kintel If the icons are ok, maybe you want to merge those into the PR, for reference.

Also, the new icons look minimally different than the old ones. If this is a problem I would suggest recreating all icons from the svg version.

File: src/export_x3d.cc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+static void append_x3d(const shared_ptr<const Geometry> &geom, std::ostream &output)
+{
+	if (const CGAL_Nef_polyhedron *N = dynamic_cast<const CGAL_Nef_polyhedron *>(geom.get())) {
+		append_x3d(*N, output);
+	}
+	else if (const PolySet *ps = dynamic_cast<const PolySet *>(geom.get())) {
+		auto r = export_indexed_triangles(*ps);
+		append_x3d(r.first, r.second, output);
+	}
+	// else if (const Polygon2d *poly = dynamic_cast<const Polygon2d *>(geom.get())) {
```
Comment (kintel) [inline src/export_x3d.cc]: This check should be re-enabled.

File: src/export_obj.cc (hunk 8 above)
Comment (kintel) [reply src/export_obj.cc]: I think this should simply be re-enabled then.

**Comment by kintel** [comment]
@donbright You worked on smth. similar a while back. I feels like this could be merged early and deal with import as a separate task. Opinions?

**Comment by kintel** [comment]
@tu500 ..and there's a merge conflict with `openscad.cc`..

**Comment by jgwinner** [comment]
I found a need for OBJ or GLTF export, and found this open ticket ...

Looks like it was a conflicted merge? What's the  status?

I was trying to build a parametric object that could be exported to GLTF or OBJ, for inclusion in a Mixed Reality Extension (Node.JS). 

        == John ==

**Comment by t-paul** [comment]
I guess the current status would be abandoned, but it might not be too complicated to revive it. As listed in the task list, some test cases would be needed too (basically integration with the existing test framework).

**Comment by kintel** [comment]
Closing as stalled. OBJ import/export has since been merged

**Pull request closed without merging**